packages/docusaurus-mdx-loader/src/remark/toc/index.ts

```diff
@@ -93,21 +94,82 @@ const plugin: Plugin = function plugin(
     const {toString} = await import('mdast-util-to-string');
     const {visit} = await import('unist-util-visit');
 
-    const headings: TOCItem[] = [];
+    const partialComponentToHeadingsName = new Map<string, string>();
 
-    visit(root, 'heading', (child: Heading) => {
-      const value = toString(child);
+    const headings: (TOCItem | NestedTOC)[] = [];
+
+    function visitHeading(node: Heading) {
+      const value = toString(node);
 
       // depth:1 headings are titles and not included in the TOC
-      if (!value || child.depth < 2) {
+      if (!value || node.depth < 2) {
         return;
       }
 
       headings.push({
-        value: toValue(child, toString),
-        id: child.data!.id!,
-        level: child.depth,
+        value: toValue(node, toString),
+        id: node.data!.id!,
+        level: node.depth,
       });
+    }
+
+    function visitMdxjsEsm(node: MdxjsEsm) {
+      if (!node.data?.estree) {
+        return;
+      }
+
+      for (const importDeclaration of node.data.estree.body) {
+        if (importDeclaration.type !== 'ImportDeclaration') {
+          continue;
+        }
+
+        const importPath = importDeclaration.source.value as string;
+        const isMdxImport = /\.mdx?$/.test(importPath);
+        if (!isMdxImport) {
+          continue;
+        }
+
+        const componentName = importDeclaration.specifiers.find(
+          (o: Node) => o.type === 'ImportDefaultSpecifier',
+        )?.local.name;
+
+        if (!componentName) {
+          continue;
+        }
+        const {size} = partialComponentToHeadingsName;
+        const exportAsName = `__${name}${size}`;
+        partialComponentToHeadingsName.set(componentName, exportAsName);
+
+        importDeclaration.specifiers.push({
+          type: 'ImportSpecifier',
+          imported: {type: 'Identifier', name},
+          local: {type: 'Identifier', name: exportAsName},
+        });
+      }
+    }
+
+    function visitMdxJsxFlowElement(node: MdxJsxFlowElement) {
+      const nodeName = node.name;
+      if (!nodeName) {
+        return;
+      }
+      const headingsName = partialComponentToHeadingsName.get(nodeName);
+      if (headingsName) {
+        headings.push({
+          nested: true,
+          name: headingsName,
+        });
+      }
+    }
+
+    visit(root, (child: Node) => {
```

Comment:
Remove `: Node` here, then you can also remove the other `as` casts.
The types for the ASTs are already quite smart, casts like `: Node` and `as Heading` should not be needed

Reply:
For me this doesn't work:

```tsx
    visit(root, (child) => {
      if (child.type === 'heading') {
        const heading: Heading = child
      } 
    });
```

How is this supposed to work?

Maybe we should ask TS to load some declaration merging typedefs containing a registry of types?